### PR TITLE
Fix backward compatibility caused by missing eval metrics class 

### PR DESCRIPTION
**Pull request by bigning**
# What does this PR do?

Fix backward compatibility. We removed some metrics classes in PR https://github.com/mosaicml/composer/pull/3243, which broke the backward compatibility for loading old checkpoints that have those metrics pickled. 
image

So we add dummy class to fix this. 

test: eval-7b-composer-15Qujf
eval-7b-composer-gGTskn

### CODE REVIEW

File: composer/metrics/nlp.py [hunk 1]
```diff
@@ -178,3 +178,26 @@ def compute(self) -> Tensor:
...
         avg_loss = super().compute()
         return torch.exp(avg_loss)
+
+
+# Fix backward compatibility
+class InContextLearningMetric:
+
+    def __init__(self):
+        raise RuntimeError(
+            f'This is a class for only fixing the backward compatibility, please use this https://github.com/mosaicml/llm-foundry/blob/main/scripts/eval/README.md'
```
Comment (dakinggg) [inline composer/metrics/nlp.py:188]:
```suggestion
            f'This only exists for maintaining backward compatibility for checkpoints that contain pickled metrics, please use this https://github.com/mosaicml/llm-foundry/blob/main/scripts/eval/README.md.'
```

**Review by dakinggg — APPROVED** [review]

File: composer/metrics/nlp.py [hunk 2]
```diff
@@ -178,3 +178,26 @@ def compute(self) -> Tensor:
...
         avg_loss = super().compute()
         return torch.exp(avg_loss)
+
+
+# Fix backward compatibility
+class InContextLearningMetric:
+
+    def __init__(self):
+        raise RuntimeError(
+            f'This only exists for maintaining backward compatibility for checkpoints that contain pickled metrics, please use this https://github.com/mosaicml/llm-foundry/blob/main/scripts/eval/README.md.'
```
Comment (dakinggg) [inline composer/metrics/nlp.py:188]:
```suggestion
            f'This class only exists for maintaining backward compatibility for checkpoints that contain pickled metrics, please use this https://github.com/mosaicml/llm-foundry/blob/main/scripts/eval/README.md.'
```

File: composer/metrics/nlp.py [hunk 3]
```diff
@@ -178,3 +178,27 @@ def compute(self) -> Tensor:
...
+class InContextLearningMetric:
+    """This only exists for backwards compatibility of checkpoints that contain pickled metrics."""
+
+    def __init__(self):
+        raise RuntimeError(
+            f'This class only exists for maintaining backward compatibility for checkpoints that contain pickled metrics, please use this https://github.com/mosaicml/llm-foundry/blob/main/scripts/eval/README.md.'
+        )
+
+    def __getstate__(self):
+        return None
```
Comment (eracah) [inline composer/metrics/nlp.py:193]: probably only need `__setstate__`, but yeah prob fine to have this

**Review by eracah — COMMENTED** [review]
nice!

File: composer/metrics/nlp.py [hunk 4]
```diff
@@ -178,3 +178,27 @@ def compute(self) -> Tensor:
         """Returns torch.exp() of the LanguageCrossEntropy."""
         avg_loss = super().compute()
         return torch.exp(avg_loss)
+
+
+# Fix backward compatibility
+class InContextLearningMetric:
+    """A class only exists for backwards compatibility of checkpoints that contain pickled metrics."""
```
Comment (mvpatel2000) [inline composer/metrics/nlp.py:185]:
```suggestion
    """InContextLearningMetric only exists for backwards compatibility of checkpoints that contain pickled metrics."""
```
nit: more concise

File: composer/metrics/nlp.py [hunk 5]
```diff
@@ -178,3 +178,27 @@ def compute(self) -> Tensor:
         """Returns torch.exp() of the LanguageCrossEntropy."""
         avg_loss = super().compute()
         return torch.exp(avg_loss)
+
+
+# Fix backward compatibility
```
Comment (mvpatel2000) [inline composer/metrics/nlp.py:183]:
```suggestion
# For backward compatibility
```

File: composer/metrics/nlp.py [hunk 6]
```diff
@@ -178,3 +178,27 @@ def compute(self) -> Tensor:
...
+class InContextLearningMetric:
+    """A class only exists for backwards compatibility of checkpoints that contain pickled metrics."""
+
+    def __init__(self):
+        raise RuntimeError(
+            f'This class only exists for maintaining backward compatibility for checkpoints that contain pickled metrics, please use this https://github.com/mosaicml/llm-foundry/blob/main/scripts/eval/README.md.',
```
Comment (mvpatel2000) [inline composer/metrics/nlp.py:189]:
```suggestion
            f'This class only exists for maintaining backward compatibility for checkpoints that contain pickled metrics. Please instead use https://github.com/mosaicml/llm-foundry/blob/main/scripts/eval/README.md.',
```

**Review by mvpatel2000 — APPROVED** [review]
LGTM! Mnor nits

**Pull request merged**